### PR TITLE
Simple variants of runT and bindT

### PR DESCRIPTION
**Pull request by KingoftheHomeless**
A modified variant of @tek's [simple `runT` and `bindT` implementation](https://github.com/tek/polysemy/commit/36325a2027c66de5eec8b1b9548eb3bd992f8645). In comparison, this version:

- Doesn't require any extra type variables
- Allows the new combinators to also be used in `reinterpretHn`, and not just in `interpretH`
- Calls the combinators `runTH` and `bindTH` instead of `interpretHO` and `interpretBindHO`.

This partially adresses #386.

These names of `runTH` and `bindTH` still aren't final unless we want them to be.

### CODE REVIEW

**Comment by isovector** [comment]
`TH` sounds like template haskell to me. What is the `H` supposed to suggest?

File: src/Polysemy/Internal/Combinators.hs [hunk 1]
```diff
@@ -166,13 +165,13 @@ reinterpretH
...
+reinterpretH f sem = Sem $ \k -> runSem sem $ \u ->
   case decompCoerce u of
     Left x  -> k $ hoist (reinterpretH f) $ x
     Right (Weaving e s d y v) -> do
-      a <- usingSem k $ runTactics s (raiseUnder . d) v $ f e
-      pure $ y a
-{-# INLINE[3] reinterpretH #-}
+      fmap y $ usingSem k
+             $ runTactics s (raiseUnder . d) v (reinterpretH f . d)
+             $ f e
```
Comment (tek) [inline src/Polysemy/Internal/Combinators.hs:174]: did you purposefully remove that inline pragma?

File: src/Polysemy/Internal/Combinators.hs [hunk 2]
```diff
@@ -80,12 +80,11 @@ interpretH
        -- already in 'Sem'.
     -> Sem (e ': r) a
     -> Sem r a
-interpretH f (Sem m) = m $ \u ->
+interpretH f (Sem m) = Sem $ \k -> m $ \u ->
```
Comment (tek) [inline src/Polysemy/Internal/Combinators.hs:83]: just out of curiosity, why did you change that?

**Comment by KingoftheHomeless** [comment]
"Higher". It's a kind of reference that the corresponding action that would be returned by `runT` is interpreted recursively and run immediately.
... Yeah, the names aren't stellar. As @tek 's pointed out, we'd ideally express it as `monadicallySequenceEmbeddedHigherOrderThunkAndRunCurrentInterpreterOnIt`, but that's just a tad too verbose.

File: src/Polysemy/Internal/Combinators.hs (hunk 1 above)
Comment (KingoftheHomeless) [reply src/Polysemy/Internal/Combinators.hs:174]: No, I've since readded that back in.

File: src/Polysemy/Internal/Combinators.hs (hunk 2 above)
Comment (KingoftheHomeless) [reply src/Polysemy/Internal/Combinators.hs:83]: Running a `Sem` as the universally quantified monad `m` inside of another `Sem` has always been more efficient than running a `Sem` as a `Sem` directly, so I figured I'd do that optimization while I was at it.

**Comment by tek** [comment]
> ... Yeah, the names aren't stellar. As @tek 's pointed out, we'd ideally express it as `monadicallySequenceEmbeddedHigherOrderThunkAndRunCurrentInterpreterOnIt`, but that's just a tad too verbose.

looking at the doc, we can even replace `Thunk` with `Continuation` to get full Bingo

File: src/Polysemy/Internal/Combinators.hs (hunk 2 above)
Comment (googleson78) [reply src/Polysemy/Internal/Combinators.hs:83]: so adding more stuff makes it go faster? :grin:

**Review by tek — APPROVED** [review]

**Comment by KingoftheHomeless** [comment]
We decided to come full circle and name them `runTSimple` and `bindTSimple` as per the original suggestions in #386.
